department-of-veterans-affairs/vets-api
# API-41099-follow-up-on-va-notify-job-status

## Summary
* Adds follow up job
* Checks statues on notification
* Uses status to re-queue or not
* Adds tests for new sidekiq follow up job
* Adds a service for the JWT encoding required for the va notify call
* REFACTOR - Removes unused line of code from VA Notify spec

* The aim of this PR is that when a VA notification is created it will get the ID of that notification now, and schedule a sidekiq job that checks to verify that it has been delivered, or reports if there is an issue.  By 'reports' I mean send s a slack notification to our alerts channel.

## Related issue(s)
[API-41099](https://jira.devops.va.gov/browse/API-41099)

#### Testing Notes
* Enable Flipper `lighthouse_claims_api_v2_poa_va_notify`

1.  Setup
    * In order for the notification to send you will need to change the `recipient_identifier` line in the `va_notify_job.rb` file so it sends to your email, otherwise it will error out and not in the way we want to track via the slack alert.  Relevant lines are 56 and 73 depending on which endpoint you are going to test.  Change to `email_address: 'you.email@here'`
2. Create a notification
    * submit a POST to the /2122 or 2122a endpoint
    * You can see the [previous PR ](https://github.com/department-of-veterans-affairs/vets-api/pull/18687)for creating the notification for further information, all the `new.perform` it not strictly required just makes stepping through each step a little easier.
    * You can verify the ID and see what the response looks like coming back from VA Notify by debugging on line 24 of the  `va_notify_job.rb` file.  It might be worth it to just copy down the ID value so you can use it for testing the slack alert later.
3. Verify its status as 'delivered'
    *  You will not see a slack message if the status comes back as `delivered` you will just see the job not queue up again (empty queue that is)
    *  You can 'fake' a failure by just hardcoding one of the other statuses into the `va_notify_follow_up_job.rb` file on line 23 for the `status`.  An example would be `sent`.  This job will re queue, that is the expected behavior, only notifying once the retries have exhausted
    * change the channel for the slack client to `vaapi-alerts-testing` on line 52 in `service_base.rb`
    * Using the Id value from the notification you can trigger the job to see the slack alert from the rails console `ClaimsApi::VANotifyFollowUpJob.perform_async('< your ID value here >')`
    * The alert will go to the `vaapi-alerts-testing` channel
    * To test the alert for when retries exhaust  update the retry count to something line 3 or 4 with a status that re-queues.

## Testing done


## Screenshots

## What areas of the site does it impact?
	new file:   modules/claims_api/app/sidekiq/claims_api/va_notify_follow_up_job.rb
	modified:   modules/claims_api/app/sidekiq/claims_api/va_notify_job.rb
	new file:   modules/claims_api/spec/sidekiq/va_notify_follow_up_job_spec.rb
	modified:   modules/claims_api/spec/sidekiq/va_notify_job_spec.rb
    new file:   modules/claims_api/app/services/claims_api/jwt_encoder.rb

## Acceptance criteria


## Requested Feedback
